requests-for-comments/2022-03-08-full-saml.md

```diff
@@ -0,0 +1,48 @@
+# RFC: Full SAML support
+
+- **Created:** *2022-03-08*
+- **Owner:** @paolodamico (and later @Twixes)
+- **GitHub issues:** [#5647](https://github.com/PostHog/posthog/issues/5647)
+
+> While commenting on this, please comment on specific lines if possible, to thread the discussion.
+
+## Overview
+
+Currently we support SAML only on self-hosted. More users have requested support for this feature on the Cloud version too. Further, the self-hosted version does not support multitenancy SAML. This RFC contains a technical proposal for how to implement full multitenant SAML support.
+
+## Background
+
+From the original SAML issue on GitHub,
+> This is a compliance feature that a lot of large-scale organizations require for full widespread rollout. When working at a large org, having a centralized place for provisioning and managing users can make the onboarding/offboarding process smoother and help reduce security risks. There's a great [Slack thread](https://posthogusers.slack.com/archives/CT7HXDEG3/p1628859965082200) by a community member about the importance of SAML and [this study](https://www.onelogin.com/press-center/press-releases/97-percent-of-saas-vendors-backing-saml-based-single-sign-on) by OneLogin in which they find that up to 97% of SaaS support have support (or plans) for SAML. 
+
+
+## Proposal
+
+1. SAML configuration becomes organization-based instead of instance-based. We create a separate model for `OrganizationConfiguration` where we track all SAML parameters (ACS URL, Entity ID, X509 certificate and attribute names [4]), and of course drop all configuration from instance environment variables.
+2. For the next release, when updating, if SAML configuration is set and the instance has a single organization, we set the same SAML parameters for that organization.
+3. Drop `SAML_ENFORCED` in favor of a `SSO_ENFORCEMENT` attribute that supports `saml`, `google-oauth2`, `github` & `gitlab` to solve [#7535](https://github.com/PostHog/posthog/issues/7535) too (only one per organization).
+4. Drop `SAML_DISABLED` support to make maintenance easier and avoid having multiple build flows. Drawback is that the FOSS version which does not support SAML will still require the SAML dependencies.
+5. If SAML is enabled for at least one organization in the instance, the login page shows only the email input at first. Upon entering an email, we do an API request as shown below (where `enforced_sso` contains information on whether to enforce log in with a particular SSO provider [either SAML, Google, ...] or `null` if the user has no SSO enforced **or the user does not exist**).
+    ```
+    GET /api/users/pre-login
+    {
+        "email": "test@posthog.com"
+    }
+
+    Response:
+    {
+        "enforced_sso": "saml"
+    }
+    ```
+
+    > ⚠️ Security concern. If the user has any SSO enforced, it will leak information about a specific user being registered in the PostHog instance.
```

Comment:
@Twixes maybe we should rate limit the API endpoint? Or is there another approach that doesn't leak info?

Reply:
Now @mariusandra

Reply:
@guidoiaquinti is the one whose input I'd listen to here. I haven't worked at a scale where this was a real threat I had to worry about. Guido, any best practices or suggestions regarding API limits from ye olde days that we could adopt as well? We are talking about a public endpoint.

Reply:
Yeah makes sense @guidoiaquinti, let's do domain-based.